### PR TITLE
Feature/ser 13,31 get,list

### PR DESCRIPTION
**Pull request by RuchitaGarde**
#### What's this PR do?
This functionality allows you to get a message from the DB. It also allows you to get a list of messages, & has the functionality to limit the messages by number, sent-by & decide whether you want to view just the summary or not. This also marks read messages as read. 
#### Related JIRA tickets:
https://jira.amida-tech.com/browse/SER-13
https://jira.amida-tech.com/browse/SER-31
#### How should this be manually tested?
- Non-manual testing: `yarn test`
- Get message (GET query): `http://localhost:4000/api/message/get/message_id`
- List messages (GET query): `http://localhost:4000/api/message/list`
- List message with parameters (GET query): `http://localhost:4000/api/message/list?limit=<no.ofMessages>&from=<username>&summary=true`
(Without the <,>)

Manually checking the DB:
- Message sent by a user should have a readAt filled in.
- Message retreived using 'get/message_id' should have readAt updated.
#### Any background context you want to provide?
A combination of one or more parameters can exist in the get query for 'list'.
For eg: `http://localhost:4000/api/message/list?summary=<true>` to get summary of all returned messages without limiting by number or sender.

A new package **chai-date-string** was added to devDependencies.
#### Screenshots (if appropriate):

### CODE REVIEW

File: server/routes/message.route.js [hunk 1]
```diff
@@ -8,11 +8,21 @@ const router = express.Router(); // eslint-disable-line new-cap
...
+ * - summary: boolean, can return a summary version of messages
+ */
+router.route('/list/')
     .get(messageCtrl.list);
-// TODO: needs to be clear that userId is the owner
-router.route('/count/:userId')
+
+/**
+ * url params:
+ * - owner: username of owner of messages to count
```
Comment (jsachs) [inline server/routes/message.route.js:22]: @RuchitaGarde why does `count` need an owner but `list` does not? Shouldn't the username/userId be coming from the JWT in both cases?

Additionally, do you want to modify this code now to user `userId` instead of `username`, or did you want to do that in a separate PR?

File: server/tests/message.integration.spec.js [hunk 2]
```diff
@@ -132,177 +145,230 @@ describe('Message API:', function () {
...
                 .expect(httpStatus.OK)
-                .then(res => {
-                    expect(res.text).to.equal('OK');
-                    done();
-                })
-                .catch(done);
+            done();
         });
 
+        // this cannot be tested correctly without auth microservice.
```
Comment (jsachs) [inline server/tests/message.integration.spec.js]: This is where I think you should generate a token from jwt.io and add it to the test harness.

File: server/tests/message.integration.spec.js [hunk 3]
```diff
@@ -132,177 +145,230 @@ describe('Message API:', function () {
...
+                    expect(res.body).to.be.an('array');  //changed res to res.body
+                    expect(res.body.length).to.equal(limit);
                     done();
                 })
                 .catch(done);
         });
 
-        it('should return a count for total Messages', done => {
+        // TODO: Ruchita to write this test
+        it('has an option to limit by sender', done => {
```
Comment (jsachs) [inline server/tests/message.integration.spec.js]: This test does not check everything. You need to make sure that you have a message from the given sender, but you also need to check *all* returned messages and make sure that there are no messages from other senders.

File: server/tests/message.integration.spec.js [hunk 4]
```diff
@@ -132,177 +145,230 @@ describe('Message API:', function () {
...
-                .get(baseURL + '/message/get' + messageId)
+                .get(baseURL + '/message/get/' + messageId)
                 .expect(httpStatus.OK)
                 .then(res => {
-                    expect(res.body).to.deep.include(testMessageObject);
+                    expect(res.body.to).to.deep.equal(testMessageObject.to);
+                    expect(res.body.from).to.equal(testMessageObject.from);
+                    expect(res.body.subject).to.equal(testMessageObject.subject);
+                    expect(res.body.message).to.equal(testMessageObject.message);
+                    //not checking for owner here because that value was forced to a string
```
Comment (jsachs) [inline server/tests/message.integration.spec.js]: I think you should still check for owner. Once you generate a JWT to use, you will get the userId/username from the token, and can check for that attribute here.

File: server/tests/message.integration.spec.js [hunk 5]
```diff
@@ -132,177 +145,230 @@ describe('Message API:', function () {
...
                     done();
                 })
                 .catch(done);
         });
 
         it('should mark the Message retrieved as read', done => {
```
Comment (jsachs) [inline server/tests/message.integration.spec.js]: This test could use another block. You should explicitly check that the message is not read initially. You could use a direct Sequelize query, then make the API call.

**Review by jsachs — CHANGES_REQUESTED** [review]
See comments. Also, submitted a few new commits with style updates.

File: server/routes/message.route.js (hunk 1 above)
Comment (RuchitaGarde) [reply server/routes/message.route.js:22]: 1) `count` does not need an owner. That is my mistake. The `owner` param needs to be removed.
2) I would like to do that in a separate PR. I don't want to mix things up right now. I will do that after this branch gets merged with develop.

**Comment by ElijahGeorge** [comment]
Closing this PR as this branch has been merged into https://github.com/amida-tech/amida-messaging-microservice/pull/14. Thanks for the great work @RuchitaGarde
